Make DeserializationException a ReusableException.

scott-gibson-sociomantic opened the pull request:
Previously the formatting call would allocate a new buffer every time the enforce triggered.

scott-gibson-sociomantic commented:
Stress testing an application revealed that even though the exception is static in `Deserializer`, the formatting calls are creating a new buffer each time.

nemanja-boric-sociomantic commented:
Alternative (and more expected IMHO) approach would be to make `DeserializationException` reusable exception: https://github.com/sociomantic-tsunami/ocean/blob/bcd88397f484ca4821ceed942b518f05c2e2274f/src/ocean/core/Exception.d#L100

mihails-strasuns-sociomantic commented:
Indeed, let's avoid repeating boilerplate.

scott-gibson-sociomantic commented:
Alright i'll update it to a ReusableException now.

scott-gibson-sociomantic committed:
Make DeserializationException a ReusableException.

Previously the formatting call would allocate a new buffer every time
the enforce triggered.

scott-gibson-sociomantic commented:
Updated to use ReusableException and also added some coverage tests.

mihails-strasuns-sociomantic committed:
Support `Buffer` in `Formatter`

mihails-strasuns-sociomantic commented:
@scott-gibson-sociomantic it never came to my mind that now we can do nice formatting of reusable messages with new formatter sink!

I took the opportunity to generalize it and move into `ReusableExceptionImpl` itself and force pushes result into your branch :P FYI @mathias-lang-sociomantic too

mathias-lang-sociomantic commented on src/ocean/core/Exception.d:
```diff
@@ -198,53 +197,40 @@ public template ReusableExceptionImplementation()
+                fmt = "{}";
+            ocean.text.convert.Formatter.sformat(this.reused_msg, fmt, arg);
+        }
+        else static if (is(T : cstring))
```
Not so fan of mixing semantics though...
`append!(cstring, cstring)` => format
`append!(uint, cstring)` => append

That is a weird API to me.

mihails-strasuns-sociomantic replied:
It was not intended as a feature - can't overload two different templated `append` methods in D1 but must preserve backwards compatibility. It is one of stupid legacy quirks that will have to wait for full D2 :(

mathias-lang-sociomantic replied:
I would much prefer to introduce a new symbol then... Maybe `format` ?

mihails-strasuns-sociomantic replied:
Sure, renamed to `fmtAppend` and restored old methods as-is. Ugly but I see your point.

mihails-strasuns-sociomantic committed:
Add ReusableExceptionImpl.fmtAppend

mihails-strasuns-sociomantic commented:
Ping @sociomantic-tsunami/core-team

nemanja-boric-sociomantic reviewed (approved):
good stuff!

nemanja-boric-sociomantic commented on src/ocean/core/Exception.d:
```diff
@@ -100,9 +100,9 @@ unittest
 public template ReusableExceptionImplementation()
 {
     import ocean.transition;
-    static import ocean.text.convert.Integer_tango;
```
is this really gone?

mihails-strasuns-sociomantic replied:
not anymore, good catch

nemanja-boric-sociomantic replied:
Good thing we don't have any unittest so this doesn't go red ;-)

mihails-strasuns-sociomantic replied:
Updated

mihails-strasuns-sociomantic replied:
> Good thing we don't have any unittest so this doesn't go red ;-)

We have but there are other imports in same module that provide it so it kept being green :(

scott-gibson-sociomantic commented:
Thanks @mihails-strasuns-sociomantic for your work. I'm unsure about the codecov/changes check but i'm happy with the new method.
